File: game/src/main/java/com/aa/mtg/game/init/test/InitTestService.java
package com.aa.mtg.game.init.test;

import com.aa.mtg.cardinstance.CardInstanceFactory;
import com.aa.mtg.cards.Card;
import com.aa.mtg.cards.Cards;
import com.aa.mtg.game.status.GameStatus;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public abstract class InitTestService {
    private static final Logger LOGGER = LoggerFactory.getLogger(InitTestService.class);

    protected CardInstanceFactory cardInstanceFactory;
    protected Cards cards;

    public void initGameStatusForTest(GameStatus gameStatus) {
        LOGGER.warn("Application is running in test mode: Initializing the gameStatus with test data.");

        // Clear
        gameStatus.getCurrentPlayer().getLibrary().getCards().clear();
        gameStatus.getCurrentPlayer().getHand().getCards().clear();
        gameStatus.getNonCurrentPlayer().getLibrary().getCards().clear();
        gameStatus.getNonCurrentPlayer().getHand().getCards().clear();

        // Call abstract initStatus
        initGameStatus(gameStatus);
    }

    public abstract void initGameStatus(GameStatus gameStatus);

    protected void addCardToCurrentPlayerLibrary(GameStatus gameStatus, Card card) {
        gameStatus.getCurrentPlayer().getLibrary().addCard(cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.getCurrentPlayer().getName()));
    }

    protected void addCardToCurrentPlayerHand(GameStatus gameStatus, Card card) {
        gameStatus.getCurrentPlayer().getHand().addCard(cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.getCurrentPlayer().getName()));
    }

    protected void addCardToCurrentPlayerBattlefield(GameStatus gameStatus, Card card) {
        String owner = gameStatus.getCurrentPlayer().getName();
        String controller = gameStatus.getCurrentPlayer().getName();
        gameStatus.getCurrentPlayer().getBattlefield().addCard(cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, owner, controller));
    }

    protected void addCardToCurrentPlayerGraveyard(GameStatus gameStatus, Card card) {
        gameStatus.getCurrentPlayer().getGraveyard().addCard(cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.getCurrentPlayer().getName()));
    }

    protected void addCardToNonCurrentPlayerLibrary(GameStatus gameStatus, Card card) {
        gameStatus.getNonCurrentPlayer().getLibrary().addCard(cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.getNonCurrentPlayer().getName()));
    }

    protected void addCardToNonCurrentPlayerHand(GameStatus gameStatus, Card card) {
        gameStatus.getNonCurrentPlayer().getHand().addCard(cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.getNonCurrentPlayer().getName()));
    }

    protected void addCardToNonCurrentPlayerBattlefield(GameStatus gameStatus, Card card) {
        String owner = gameStatus.getNonCurrentPlayer().getName();
        String controller = gameStatus.getNonCurrentPlayer().getName();
        gameStatus.getNonCurrentPlayer().getBattlefield().addCard(cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, owner, controller));
    }

    protected void addCardToNonCurrentPlayerGraveyard(GameStatus gameStatus, Card card) {
        gameStatus.getNonCurrentPlayer().getGraveyard().addCard(cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.getNonCurrentPlayer().getName()));
    }

    public void setCardInstanceFactory(CardInstanceFactory cardInstanceFactory) {
        this.cardInstanceFactory = cardInstanceFactory;
    }

    public void setCards(Cards cards) {
        this.cards = cards;
    }

    public CardInstanceFactory getCardInstanceFactory() {
        return cardInstanceFactory;
    }

    public Cards getCards() {
        return cards;
    }
}

File: src/test/kotlin/application/InitTestService.kt
package application

import com.matag.cards.Card
import com.matag.cards.Cards
import com.matag.game.cardinstance.CardInstanceFactory
import com.matag.game.init.InitService
import com.matag.game.status.GameStatus

abstract class InitTestService(
    protected val cardInstanceFactory: CardInstanceFactory,
    protected val cards: Cards
) : InitService {

    abstract fun initTestGameStatus(gameStatus: GameStatus)

    override fun initGameStatus(gameStatus: GameStatus) {
        gameStatus.currentPlayer.hand.extractAllCards()
        gameStatus.nonCurrentPlayer.hand.extractAllCards()
        initTestGameStatus(gameStatus)
    }

    protected fun addCardToCurrentPlayerLibrary(gameStatus: GameStatus, card: Card) {
        gameStatus.currentPlayer.library.addCard(
            cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.currentPlayer.name)
        )
    }

    protected fun addCardToCurrentPlayerHand(gameStatus: GameStatus, card: Card) {
        gameStatus.currentPlayer.hand.addCard(
            cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.currentPlayer.name)
        )
    }

    protected fun addCardToCurrentPlayerBattlefield(gameStatus: GameStatus, card: Card) {
        val owner = gameStatus.currentPlayer.name
        val controller = gameStatus.currentPlayer.name
        gameStatus.currentPlayer.battlefield.addCard(
            cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, owner, controller)
        )
    }

    protected fun addCardToCurrentPlayerGraveyard(gameStatus: GameStatus, card: Card) {
        gameStatus.currentPlayer.graveyard.addCard(
            cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.currentPlayer.name)
        )
    }

    protected fun addCardToNonCurrentPlayerLibrary(gameStatus: GameStatus, card: Card) {
        gameStatus.nonCurrentPlayer.library.addCard(
            cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.nonCurrentPlayer.name)
        )
    }

    protected fun addCardToNonCurrentPlayerHand(gameStatus: GameStatus, card: Card) {
        gameStatus.nonCurrentPlayer.hand.addCard(
            cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.nonCurrentPlayer.name)
        )
    }

    protected fun addCardToNonCurrentPlayerBattlefield(gameStatus: GameStatus, card: Card) {
        val owner = gameStatus.nonCurrentPlayer.name
        val controller = gameStatus.nonCurrentPlayer.name
        gameStatus.nonCurrentPlayer.battlefield.addCard(
            cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, owner, controller)
        )
    }

    protected fun addCardToNonCurrentPlayerGraveyard(gameStatus: GameStatus, card: Card) {
        gameStatus.nonCurrentPlayer.graveyard.addCard(
            cardInstanceFactory.create(gameStatus, gameStatus.nextCardId(), card, gameStatus.nonCurrentPlayer.name)
        )
    }
}
